client/allocdir/fs_linux.go

```diff
@@ -25,19 +24,54 @@ func linkDir(src, dst string) error {
 		return err
 	}
 
-	return syscall.Mount(src, dst, "", syscall.MS_BIND, "")
+	return unix.Mount(src, dst, "", unix.MS_BIND, "")
 }
 
 // unlinkDir unmounts a bind mounted directory as Linux doesn't support
 // hardlinking directories. If the dir is already unmounted no error is
 // returned.
 func unlinkDir(dir string) error {
-	if err := syscall.Unmount(dir, 0); err != nil {
-		if err != syscall.EINVAL {
+	return unmount(dir)
+}
+
+// bindMount mounts src to dst with support for read only mounting.
+// Assumes that filepath.Dir(dst) exists already
+func bindMount(src, dst string, readOnly bool) error {
+	if _, err := os.Stat(dst); os.IsNotExist(err) {
+		// check if destination is available and create a corresponding type if it is not present
+		stats, err := os.Stat(src)
+		if err != nil {
 			return err
 		}
+
+		if stats.IsDir() {
+			if err := os.Mkdir(dst, 0777); err != nil {
+				return err
+			}
+		} else {
+			if _, err := os.Create(dst); err != nil {
+				return err
+			}
+
+		}
+	}
+
+	err := unix.Mount(src, dst, "", unix.MS_BIND, "")
+	if err != nil || !readOnly {
+		return err
+	}
+
+	// mount yet again for read-only flag
+	return unix.Mount("", dst, "", unix.MS_BIND|unix.MS_RDONLY|unix.MS_REMOUNT, "")
```

Comment:
I could not find a way to make this an atomic operation - I inspected mount behavior and found the following:

```
$ sudo strace -o mount.trace mount --read-only --bind ./source ./target
$ grep 'mount(' mount.trace
mount("/tmp/playground/source", "/tmp/playground/target", 0x1493270, MS_MGC_VAL|MS_RDONLY|MS_BIND, NULL) = 0
mount("none", "/tmp/playground/target", NULL, MS_RDONLY|MS_REMOUNT|MS_BIND, NULL) = 0
```